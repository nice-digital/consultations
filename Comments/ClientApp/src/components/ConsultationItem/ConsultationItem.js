// @flow

import React, { Component, Fragment } from "react";
import Moment from "react-moment";
import { Link } from "react-router-dom";

import { GenerateCode } from "../GenerateCode/GenerateCode";

type StateType = {}

type PropsType = {
	title: string,
	startDate: Date,
	endDate: Date,
	submissionCount: number,
	consultationId: number,
	documentId: number,
	chapterSlug: string,
	gidReference: string,
	productTypeName: string,
	hasCurrentUserEnteredCommentsOrAnsweredQuestions: boolean,
	hasCurrentUserSubmittedCommentsOrAnswers: boolean,
	isOpen: boolean,
	isClosed: boolean,
	isUpcoming: boolean,
	organisationCodes: Array,
	showShareWithOrganisationButton: boolean,
	show: boolean,
	basename: string,
	submissionToLeadCount: number,
	hidden: boolean
}

export class ConsultationItem extends Component<PropsType, StateType> {

	render() {

		const {
			title,
			startDate,
			endDate,
			submissionCount,
			consultationId,
			documentId,
			chapterSlug,
			gidReference,
			productTypeName,
			hasCurrentUserEnteredCommentsOrAnsweredQuestions,
			hasCurrentUserSubmittedCommentsOrAnswers,
			isOpen,
			isClosed,
			isUpcoming,
			organisationCodes,
			showShareWithOrganisationButton,
			submissionToLeadCount,
			hidden,
		} = this.props;

		const status = (isOpen, isClosed, isUpcoming) => {
			if (isOpen) return "Open";
			if (isClosed) return "Closed";
			if (isUpcoming) return "Upcoming";
			return "?";
		};

		const userHasRespondedButNotSubmitted = hasCurrentUserEnteredCommentsOrAnsweredQuestions & !hasCurrentUserSubmittedCommentsOrAnswers;
		const consultationStatus = status(isOpen, isClosed, isUpcoming);

		return (
			<li className="ConsultationItem">
				<article className="card">
					<header className="card__header">
						<h3 className="card__heading">
							{
								documentId != null && chapterSlug != null ?
									<Link to={`/${consultationId}/${documentId}/${chapterSlug}`}>{title}</Link>
									: title
							}
						</h3>
					</header>
					<dl className="card__metadata">
						<div className="card__metadatum">
							<dt className="visually-hidden">Consultation state</dt>
							<dd>
								<span className={`tag tag--${consultationStatus.toLowerCase()}`}>{consultationStatus}</span>
							</dd>
						</div>
						{Boolean(hidden) && (
							<div className="card__metadatum">
								<dt className="visually-hidden">Consultation state</dt>
								<dd>
									<span className={"tag tag--hidden" }>Hidden</span>
								</dd>
							</div>
						)}
						{Boolean(userHasRespondedButNotSubmitted) && (
							<div className="card__metadatum">
								<dt className="visually-hidden">Unsubmitted questions or answers</dt>
								<dd>
									<span className="tag tag--unsubmitted">Unsubmitted</span>
								</dd>
							</div>
						)}
						{showShareWithOrganisationButton && submissionToLeadCount !== null &&
							<div className="card__metadatum">
								{submissionToLeadCount} {submissionToLeadCount === 1 ? "response" : "responses"} from your organisation
							</div>
						}
						<div>
							<div className="card__metadatum">
								<dt className="visually-hidden">Project ID</dt>
								<dd title={`Consultation ID ${consultationId}`}>
									{gidReference}
								</dd>
							</div>
							<div className="card__metadatum">
								<dt className="visually-hidden">Product type name</dt>
								<dd>
									{productTypeName}
								</dd>
							</div>
							<div className="card__metadatum">
								<dd>
									{consultationStatus === "Upcoming" ?
										<Fragment>
											Starts on{" "}
											<strong><Moment format="D MMMM YYYY" date={startDate}/></strong>
										</Fragment>
										:
										<Fragment>
											Closes on{" "}
											<strong><Moment format="D MMMM YYYY" date={endDate}/></strong>
										</Fragment>
									}
								</dd>
							</div>
							{((consultationStatus !== "Upcoming") && (submissionCount !== null)) &&
								<div className="card__metadatum">
									<dd>
										{submissionCount > 0 ?
											<a href={`${this.props.basename}/api/Export/${this.props.consultationId}`} target="_blank" rel="noopener noreferrer">
												Download <strong>{submissionCount}</strong> response{submissionCount > 1 ? "s" : ""}
											</a>
											:
											<span>No responses</span>
										}
									</dd>
								</div>
							}
						</div>
					</dl>
					{showShareWithOrganisationButton &&
						<GenerateCode organisationCodes={organisationCodes} consultationId={consultationId} />
					}
				</article>
			</li>
		);
	}
}

export default ConsultationItem;
