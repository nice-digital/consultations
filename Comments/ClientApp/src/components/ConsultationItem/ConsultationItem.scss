@use "@nice-digital/nds-core/scss/colours";
@use "../../styles/variables" as *;

.ConsultationItem {
  .tag--open {
    background-color: $ec-green;
  }
  .tag--upcoming {
    background-color: $ec-orange;
  }
  .tag--unsubmitted {
    background-color: $ec-pink;
  }

  .tag--hidden {
    background-color: #666666;
    color: #FFFFFF;
  }

  .card {
    overflow: visible;

    &__heading a {
      &:before {
        display: none;
      }
    }
  }
}
