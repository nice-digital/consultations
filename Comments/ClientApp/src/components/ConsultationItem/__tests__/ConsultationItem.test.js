import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { ConsultationItem } from "../ConsultationItem";

test("does not render link if document id or chapter slug is null", async () => {
	window.__PRELOADED__ = { isAuthorised: true};
	const fakeProps = {
		title:"Consultation Title",
		startDate : new Date("1995-12-17T03:24:00"),
		endDate : new Date("2059-12-17T03:24:00"),
		submissionCount:1,
		consultationId :123,
		documentId:1,
		chapterSlug :"introduction",
		gidReference :"TA-123",
		productTypeName :"TA",
		isOpen: true,
		isClosed: false,
		isUpcoming: false,
		show: true,
		basename: "fdfd",
		showShareWithOrganisationButton: true,
		allowGenerateOrganisationCode: true,
		submissionToLeadCount: 15,
	};
	render(<MemoryRouter><ConsultationItem {...fakeProps} chapterSlug={null} documentId={null} /></MemoryRouter>);
	const consultationLink = screen.queryAllByText(fakeProps.title, { selector: "a" });
  	expect(consultationLink.length).toEqual(0);
});

test("should match snapshot with supplied data", () => {
	window.__PRELOADED__ = { isAuthorised: true};
	const fakeProps = {
		title:"Consultation Title",
		startDate : new Date("1995-12-17T03:24:00"),
		endDate : new Date("2059-12-17T03:24:00"),
		submissionCount:1,
		consultationId :123,
		documentId:1,
		chapterSlug :"introduction",
		gidReference :"TA-123",
		productTypeName :"TA",
		isOpen: true,
		isClosed: false,
		isUpcoming: false,
		show: true,
		basename: "fdfd",
		showShareWithOrganisationButton: true,
		allowGenerateOrganisationCode: true,
		submissionToLeadCount: 15,
	};
	const {container} = render(<MemoryRouter><ConsultationItem {...fakeProps} /></MemoryRouter>);
	expect(container).toMatchSnapshot();
});

