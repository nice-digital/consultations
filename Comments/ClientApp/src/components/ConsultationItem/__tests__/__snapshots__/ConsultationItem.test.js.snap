// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with supplied data 1`] = `
<div>
  <li
    class="ConsultationItem"
  >
    <article
      class="card"
    >
      <header
        class="card__header"
      >
        <h3
          class="card__heading"
        >
          <a
            href="/123/1/introduction"
          >
            Consultation Title
          </a>
        </h3>
      </header>
      <dl
        class="card__metadata"
      >
        <div
          class="card__metadatum"
        >
          <dt
            class="visually-hidden"
          >
            Consultation state
          </dt>
          <dd>
            <span
              class="tag tag--open"
            >
              Open
            </span>
          </dd>
        </div>
        <div
          class="card__metadatum"
        >
          15
           
          responses
           from your organisation
        </div>
        <div>
          <div
            class="card__metadatum"
          >
            <dt
              class="visually-hidden"
            >
              Project ID
            </dt>
            <dd
              title="Consultation ID 123"
            >
              TA-123
            </dd>
          </div>
          <div
            class="card__metadatum"
          >
            <dt
              class="visually-hidden"
            >
              Product type name
            </dt>
            <dd>
              TA
            </dd>
          </div>
          <div
            class="card__metadatum"
          >
            <dd>
              Closes on
               
              <strong>
                <time
                  datetime="2838857040000"
                >
                  17 December 2059
                </time>
              </strong>
            </dd>
          </div>
          <div
            class="card__metadatum"
          >
            <dd>
              <a
                href="fdfd/api/Export/123"
                rel="noopener noreferrer"
                target="_blank"
              >
                Download 
                <strong>
                  1
                </strong>
                 response
                
              </a>
            </dd>
          </div>
        </div>
      </dl>
      <div>
        <button
          aria-controls="organisation-codes-123"
          aria-expanded="false"
          class="btn btn--inverse"
          data-qa-sel="share-with-org-button"
          id="share-organisation-123"
          type="button"
        >
          Share with organisation 
          <span
            class="arrow-icon"
          />
        </button>
      </div>
    </article>
  </li>
</div>
`;
